### DISCUSSION TITLE
Backup Facility

### DISCUSSION THREAD

**Discussion by SomeGeek** [modoboa/modoboa#3862] [Ideas]
Modo should have a backup facility for both administrators and users.

This backup can contain:
- User Mailboxes
- User Configurations
- Applied Filters

And anything else I might have not yet thought of.

Furthermore a scheduled backup and remote backup location would be very welcome. (SFTP/FTPS/Rsync)

**Comment by almereyda** [comment]
In general I see two kinds of backup here:

* full server backups, able to restore a whole service [probably from `-installer` CLI]
* individual backups of configuration, user data (dovecot, radicale, sieve, ...), [probably from within the WebUI]

**Comment by hadifarnoud** [comment]
@almereyda I think in my case full server backup is very urgent. individual backups can be very useful but not so important for me

**Comment by jaesivsm** [comment]
I have an ansible script that'll backup databases and user mail : https://galaxy.ansible.com/jaesivsm/modoboa-ansible-backup/
Needs improvement though since I tested restoring backup and : 
 * there's no command to push a backup
 * I had to meddle with SQL since SQL dump are full database dump and not "data only"
In the end I think I should be included in the installer

**Comment by hadifarnoud** [comment]
that's a great start @jaesivsm. I still think Modoboa should have that option but sounds like this Ansible script will be quicker way of getting it done

**Comment by hadifarnoud** [comment]
@jaesivsm can you please provide an example? I did install ansible, then get the role from ansible-galaxy

afterwards, created this yml
```
- hosts: all
  roles:
    - jaesivsm.modoboa-ansible-backup
```

I get this error when running it via `ansible-playbook modoboa-backup.yml`
```
 [WARNING]: provided hosts list is empty, only localhost is available

ERROR! no action detected in task

The error appears to have been in '/etc/ansible/roles/jaesivsm.modoboa-ansible-backup/tasks/main.yml': line 25, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- archive:
  ^ here
```

**Comment by hadifarnoud** [comment]
would be nice to have a restore manual as well :)

**Comment by hadifarnoud** [comment]
I think having a proper backup/restore functionality is a core feature. would be best to have it inside Modoboa

**Comment by avsdev-cw** [comment]
This would be a very, very good feature to add. The ansible script is nice, but there's so much it excludes.

Another option for me is to have to ability to configure a "data" directory during install which I could dump onto a partition for mirror backups. This would ensure the image could be rebuilt (ie from installer or cold store OS backup) and the data partition re-mounted/re-mirrored and off you go. There's quite a bit of work involved in this though, and a lot of discussion around what is cold store backup and hot store backup, ie. ssl certs, config files, database data, maildir data.

**Comment by calvez** [comment]
this looks smart:
`https://github.com/tachtler/dovecot-backup/blob/master/dovecot_backup.sh`
